lib/active_model/serializer/concerns/associations.rb

```diff
@@ -85,18 +85,26 @@ def associate(reflection)
       #
       def associations(include_directive = ActiveModelSerializers.default_include_directive, include_slice = nil)
         include_slice ||= include_directive
-        return unless object
+        return Enumerator.new unless object
 
         Enumerator.new do |y|
-          self.class._reflections.values.each do |reflection|
+          self.class._reflections.each do |key, reflection|
             next if reflection.excluded?(self)
-            key = reflection.options.fetch(:key, reflection.name)
             next unless include_directive.key?(key)
 
-            y.yield reflection.build_association(self, instance_options, include_slice)
+            y.yield reflect_association(reflection, key: key, include_slice: include_slice)
           end
         end
       end
+
+      # @param reflection [Reflection]
+      # @param association_options [Hash] required: include_slice
+      # @return [Association]
+      def reflect_association(reflection, association_options)
```

Comment:
Rails `concerns`/mixins/inheritance should be avoided in favour of composition. Mixins seem attractive at first but just increase the surface area of objects, making it virtually impossible to maintain them when it get to a point. With composition you keep logic in different places and you decrease cognitive load.